### PR TITLE
feat: add codemod

### PR DESCRIPTION
**Pull request by sashuk**
[FX-4260]

### Description

This pull request adds a codemod that replaces BASE-aligned spacing property values with new spacing constants.

`jscodeshift` has the following problem when running this codemod:
- empty line is inserted before the import that is added via `insertAfter()`, more in https://github.com/facebook/jscodeshift/issues/185
- the semicolon is inserted after the import that is added via `insertAfter()`, more in https://github.com/benjamn/recast/issues/226

However, both issues are solved by running lint or prettier after the codemod run (this recommendation was added to the changeset)

### How to test

- Run tests – check out this branch locally and run `yarn run jest packages/picasso-codemod/src/v38.1.0/__tests__`
- Run actual codemod in Staff Portal – check out https://github.com/toptal/staff-portal/pull/10942 branch locally, drop last commit, and run `npx @toptal/picasso-codemod v38.1.0`

### Screenshots

Typical change from https://github.com/toptal/staff-portal/pull/10942/files

Screenshot 2023-09-01 at 12 37 38

---

Running in Staff Portal (https://github.com/toptal/staff-portal/pull/10942)

Screenshot 2023-09-01 at 09 18 53

### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [N/A] Make sure that additions and changes on the design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [N/A] Annotate all `props` in component with documentation
- [N/A] Create `examples` for component
- [N/A] Ensure that deployed demo has expected results and good examples
- [N/A] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [x] Self reviewed
- [x] Covered with tests ([visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md) included)

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-4260]: https://toptal-core.atlassian.net/browse/FX-4260?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sashuk** [comment]
@toptal-bot run package:alpha-release

**Comment by sashuk** [comment]
@toptal-bot run package:alpha-release

**Comment by sashuk** [comment]
@toptal-bot run package:alpha-release

**Comment by sashuk** [comment]
@toptal-bot run package:alpha-release

**Comment by dmaklygin** [comment]
I tried to run it inside a Picasso and this is an output:

```
No files selected, nothing to do. 
All done. 
Results: 
0 errors
0 unmodified
0 skipped
0 ok
Time elapsed: 0.002seconds 
```

I think, we also should cover spec, test and story files, because they are also part of the project.

**Comment by sashuk** [comment]
@dmaklygin Great suggestion, please let me add it!

File: packages/picasso-codemod/src/v38.1.0/utils/get-updated-node.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import type { TransformationOptions } from '../types'
+import { getNodeForNumber } from './get-node-for-number'
+import { getNodeForSizeStringConstant } from './get-node-for-size-string-constant'
+
+export const getUpdatedNode = (node: any, options: TransformationOptions) => {
```
Comment (augustobmoura) [inline packages/picasso-codemod/src/v38.1.0/utils/get-updated-node.ts:5]: Maybe we should type node here with the correct JSCodeshift type, wdyt? JSCodeshift types can be a bit tricky and having them fully typed help catch some edge cases

**Comment by augustobmoura** [comment]
@toptal-bot run package:alpha-release

File: .changeset/good-houses-end.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@toptal/picasso-codemod': minor
+---
+
+- add `spacing-values` codemod. Please run the codemod (`npx @toptal/picasso-codemod v38.1.0`) to replace spacing property values of `Container` and `Dropdown` components with BASE-aligned property values according to the https://github.com/toptal/picasso/blob/master/docs/decisions/18-spacings.md. Property values that do not have BASE counterpart or are complex expressions have to be updated manually (non-BASE values have to be replaced with BASE ones after consulting with Design Team), codemod outputs the list of such cases for convenience. Run linter or prettier to align updated code with project code style.
```
Comment (augustobmoura) [inline .changeset/good-houses-end.md:5]: I would just document here to use the latest version of picasso-codemod available
```suggestion
- add `spacing-values` codemod. Please run the codemod (`npx @toptal/picasso-codemod@latest v38.1.0`) to replace spacing property values of `Container` and `Dropdown` components with BASE-aligned property values according to the https://github.com/toptal/picasso/blob/master/docs/decisions/18-spacings.md. Property values that do not have BASE counterpart or are complex expressions have to be updated manually (non-BASE values have to be replaced with BASE ones after consulting with Design Team), codemod outputs the list of such cases for convenience. Run linter or prettier to align updated code with project code style.
```

**Review by augustobmoura — APPROVED** [review]
Seems to be working as expected, but we have a bullet on AC for marking the old props as deprecated.

> Make number and SpacingEnums as deprecated

Maybe we should move this AC to https://toptal-core.atlassian.net/browse/FX-4315?

**Comment by sashuk** [comment]
@dmaklygin Just checked it in `picasso`, I think the issue was that you need to use the following command (otherwise `picasso-codemod` looks for monorepo)

```
npx @toptal/picasso-codemod v38.1.0 packages/picasso/src/**/*.ts*
```

The output is the following

```
All done. 
Results: 
1 errors
2011 unmodified
0 skipped
0 ok
Time elapsed: 3.879seconds 
```

And 1 error stands for custom case that needs to be addressed manually

```
Manual update required for Container.bottom in packages/picasso/src/Pagination/story/Ellipsis.example.tsx:79
 ERR packages/picasso/src/PageTopBarMenu/PageTopBarMenu.tsx Transformation error (Unable to find @toptal/picasso or Container declaration to insert new spacing import)
Error: Unable to find @toptal/picasso or Container declaration to insert new spacing import
    at insertSpacingImport (/Users/aleksandr/toptal/tmp/picasso/packages/picasso-codemod/src/v38.1.0/utils/insert-spacing-import.ts:28:11)
    at transform (/Users/aleksandr/toptal/tmp/picasso/packages/picasso-codemod/src/v38.1.0/spacing-values.ts:108:24)
Manual update required for Container.bottom in packages/picasso/src/Pagination/story/Variants.example.tsx:8
```

Then `yarn start` is run and Picasso Storybook launches

Screenshot 2023-09-05 at 12 41 12

---

However, one bug was caught and addressed, thanks for checking on `picasso`!

**Comment by sashuk** [comment]
@augustobmoura 

> Maybe we should move this AC to https://toptal-core.atlassian.net/browse/FX-4315?

It looks like it belongs to this ticket, I'll address it in a separate pull request, thanks!

File: packages/picasso-codemod/src/v38.1.0/utils/get-updated-node.ts (hunk 1 above)
Comment (sashuk) [reply packages/picasso-codemod/src/v38.1.0/utils/get-updated-node.ts]: @augustobmoura Thanks for spotting, proper typing was added!

**Comment by sashuk** [comment]
@toptal-bot run package:alpha-release

**Comment by sashuk** [comment]
Latest package Staff Portal run result ✅ 

Screenshot 2023-09-05 at 13 46 47

**Comment by sashuk** [comment]
@toptal-anvil ping reviewers

**Review by dmaklygin — APPROVED** [review]

**Pull request merged**